### PR TITLE
hunspell: introduce FragmentChecker to speed up ModifyingSuggester

### PR DESCRIPTION
**Pull request by donnerpeter**
add NGramFragmentChecker to quickly check whether insertions/replacements produce strings that are even possible in the language

### CODE REVIEW

File: lucene/analysis/common/src/java/org/apache/lucene/analysis/hunspell/FragmentChecker.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.lucene.analysis.hunspell;
+
+/**
+ * An oracle for quickly checking that a specific part of a word can never be a valid word. This
+ * allows speeding up the "Modification" part of {@link Suggester} but avoiding expensive checks on
```
Comment (dweiss) [inline lucene/analysis/common/src/java/org/apache/lucene/analysis/hunspell/FragmentChecker.java:21]:
```suggestion
 * allows speeding up the "Modification" part of {@link Suggester} by avoiding expensive checks on
```

File: lucene/analysis/common/src/java/org/apache/lucene/analysis/hunspell/FragmentChecker.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+ * impossible words. Implementations may use character case, n-grams or whatever they like.
+ *
+ * @see NGramFragmentChecker
+ */
+public interface FragmentChecker {
+  FragmentChecker EVERYTHING_POSSIBLE = (word, start, end) -> false;
+
+  /**
+   * Check if some substring of the word that intersects with the given offsets is impossible in the
```
Comment (dweiss) [inline lucene/analysis/common/src/java/org/apache/lucene/analysis/hunspell/FragmentChecker.java:30]:
```suggestion
   * Check if the substring of the word between {@code start} (inclusive) and {@code end} (exclusive) is impossible in the
```
I didn't check whether the inclusive/exclusive part is right but I'd use the same contract as in String substring/subsequence.

File: lucene/analysis/common/src/java/org/apache/lucene/analysis/hunspell/NGramFragmentChecker.java [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+   * latter case, consider using {@link #fromWords}.
+   *
+   * @param n the length of n-grams
+   * @param dictionary the dictionary to traverse
+   * @param checkCanceled an object that's periodically called, allowing to interrupt the traversal
+   *     by throwing an exception
+   */
+  public static NGramFragmentChecker fromAllSimpleWords(
+      int n, Dictionary dictionary, Runnable checkCanceled) {
+    BitSet hashes = new BitSet(1 << (7 + n * 3));
```
Comment (dweiss) [inline lucene/analysis/common/src/java/org/apache/lucene/analysis/hunspell/NGramFragmentChecker.java:88]: That sparsity assertion inside NGramFragment and this code makes me wonder whether this is something that you came up with experimentally or is this something you can give a pointer/reference to? :) 

I'd also restrict the maximum value of n on input (and in the docs); cap it to 5 or something - I don't think it makes sense to create a larger bitset with this implementation?

Also, it's basically a bloom filter backed by a single hash function and the collision probability decreased by making the hash function's target table large enough, am I right?...

**Review by dweiss — APPROVED** [review]
Very interesting. LGTM, with minor suggestions inlined.

File: lucene/analysis/common/src/java/org/apache/lucene/analysis/hunspell/FragmentChecker.java (hunk 2 above)
Comment (donnerpeter) [reply lucene/analysis/common/src/java/org/apache/lucene/analysis/hunspell/FragmentChecker.java:30]: Mostly it's a single-char interval there now. And the semantics is as described (albeit probably not very clearly): if there is any impossible substring (ngram) intersecting that 1-char interval, return true. Any ideas on the wording improvement?

File: lucene/analysis/common/src/java/org/apache/lucene/analysis/hunspell/NGramFragmentChecker.java (hunk 3 above)
Comment (donnerpeter) [reply lucene/analysis/common/src/java/org/apache/lucene/analysis/hunspell/NGramFragmentChecker.java:88]: It's from the experiment. I've forgot to restrict it, thanks for reminding! And yes, it's currently a single-function Bloom filter, the implementation can be changed later.

File: lucene/analysis/common/src/java/org/apache/lucene/analysis/hunspell/FragmentChecker.java (hunk 2 above)
Comment (dweiss) [reply lucene/analysis/common/src/java/org/apache/lucene/analysis/hunspell/FragmentChecker.java:30]: Ok, please disregard the suggestion. I get it now, although perhaps the wording could be improved indeed. Or maybe a concrete example based on the ngram fragment checker would clarify things more than a formal contract?

Comment (donnerpeter) [reply lucene/analysis/common/src/java/org/apache/lucene/analysis/hunspell/FragmentChecker.java]: Expanded the documentation. Hopefully it's clearer now.

**Review by dweiss — APPROVED** [review]
Thanks, Peter.

File: lucene/analysis/common/src/java/org/apache/lucene/analysis/hunspell/FragmentChecker.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+package org.apache.lucene.analysis.hunspell;
+
+/**
+ * An oracle for quickly checking that a specific part of a word can never be a valid word. This
+ * allows speeding up the "Modification" part of {@link Suggester} by avoiding expensive checks on
+ * impossible words. Implementations may use character case, n-grams or whatever they like.
+ *
+ * @see NGramFragmentChecker
+ */
+public interface FragmentChecker {
```
Comment (uschindler) [inline lucene/analysis/common/src/java/org/apache/lucene/analysis/hunspell/FragmentChecker.java:26]: I would mark this as `@FunctionalInterface`

Comment (dweiss) [reply lucene/analysis/common/src/java/org/apache/lucene/analysis/hunspell/FragmentChecker.java:26]: I never understood the intent of this annotation... What difference does it make in practice?

**Pull request merged**